### PR TITLE
Added scope logic

### PR DESCRIPTION
**Pull request by johndanenRV**

### CODE REVIEW

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (mejohnnaylor) [inline sdks/cpp/common/examples/full_service/full_service.cpp:178]: authzProcessor object goes out of scope line 179. investigate what SetAuthMetadataProcessor does with it - if it makes a copy, that's ok, if it's storing a reference, or pointer, that's a problem because it's now a dangling reference (i.e. a reference to a deleted object)

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (mejohnnaylor) [inline sdks/cpp/common/examples/full_service/full_service.cpp:258]: so we're using picojson? How is it included? What are it's license terms?

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (mejohnnaylor) [inline sdks/cpp/common/examples/full_service/full_service.cpp:324]: check this doesn't break the fix for failing to terminate server when a client is connected

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (mejohnnaylor) [inline sdks/cpp/common/examples/full_service/full_service.cpp:326]: it's understandable why you made client scopes a vector of strings. however, it could be better / more efficient to code them as a bit-field. with monitor being bits 0 & 1, operate 2 & 3 and so on.

operate read access is signaled by bit 2 being 1, no read access signaled by it being 0, write access is granted when bit 3 is 1, and prohibited when it is zero.

don't do this now - until it's proven that hauling strings around and parsing them for every parameter access is unduly burdensome - "premature optimization" is a curse!

File: sdks/cpp/common/include/ParamAccessor.h
Comment (mejohnnaylor) [inline sdks/cpp/common/include/ParamAccessor.h:45]: hey! 1970's code!
in C++ we prefer to write const std::string kAuthzDisabled("__AUTHZ_DISABLED__"); 
why?
that's a good research topic for you. something to do with being able to pass by reference, take addresses etc.

**Review by mejohnnaylor — COMMENTED** [review]
please address inline comments.

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (johndanenRV) [reply sdks/cpp/common/examples/full_service/full_service.cpp:342]: This branch was made before that fix was merged so this would not have affected it. I updated the branch and checked to be safe, and it is fine.

File: sdks/cpp/common/include/ParamAccessor.h
Comment (johndanenRV) [reply sdks/cpp/common/include/ParamAccessor.h]: Fixed

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (johndanenRV) [reply sdks/cpp/common/examples/full_service/full_service.cpp:182]: authzProcessor is a std::shared_ptr passed by reference to SetAuthMetaDataProcessor(). It is hard to know exactly what grpc does with it behind the scene but I would assume that it makes some kind of copy of the shared_ptr. Given that the AuthMetaDataProcessor is running as expected I don't think this is an issue.

Comment (mejohnnaylor) [reply sdks/cpp/common/examples/full_service/full_service.cpp:182]: yep - if it's anything like std::shared_ptr, it'll maintain a reference count and only delete the object once the final owning object is destroyed. This code is now considered safe ;-) Thanks for checking!

File: sdks/cpp/common/examples/full_service/full_service.cpp
Comment (mejohnnaylor) [reply sdks/cpp/common/examples/full_service/full_service.cpp:272]: yep, BSD 2-clause is fine. thanks!

**Review by mejohnnaylor — APPROVED** [review]

**Pull request merged**
